[NavigationDrawer] Added ability to account for bottom safe area in presentation of initial drawer height

jakerockland opened the pull request:
Added ability to account for bottom safe area in presentation of initial drawer height, parameterized by default-off flag.

Fixes #9322.

I tried adding unit tests on this, per our conversation @yarneo, but ran into some issues with writing tests around this as: 
1) unit testing for this is difficult as one cannot manually set the `safeAreaInsets` 
2) snapshot testing this requires that the test suite is done on an iPhone X or similar device with bottom insets, which is not the current test device for snapshots as far as I can tell 

I'm happy to add tests but need a bit of guidance regarding how to best test this given the above constraints. I didn't see a test testing the related `shouldIncludeSafeAreaInContentHeight` behavior, which I was looking to use as a test reference, but perhaps could do something similar to: https://github.com/material-components/material-components-ios/blob/b8090cb6382910cc4190c7030739e145f3856929/components/FlexibleHeader/tests/unit/supplemental/FlexibleHeaderTopSafeAreaTestsFakeViewController.m#L90-L94.

Wanted to get your thoughts on this before diving further @yarneo.

jakerockland committed:
Added ability to account for bottom safe area in presentation of initial drawer height, parameterized by default-off flag.

jverkoey commented:
b/146566784 maps to https://github.com/material-components/material-components-ios/issues/7169

jverkoey commented:
I've fixed that; it looks like the internal issue may have been copied from another issue resulting in two issues pointing to the same GitHub issue.

jakerockland commented:
@jverkoey sorry about that, I think I did use a different issue as a template and must not have cleared out the Github field 😓

jverkoey commented:
No worries! That field really just shouldn't be getting copied when an issue is used as a template.

yarneo commented:
Thanks @jakerockland ! Will take a look today

yarneo reviewed (commented):
Thank you for contributing!

I looked at the PR and it looks great overall.

Regarding testing the bottom safe area, i wonder if this would help: https://github.com/material-components/material-components-ios/blob/develop/components/FlexibleHeader/tests/unit/supplemental/FlexibleHeaderTopSafeAreaTestsFakeViewController.m#L90

Basically overriding the safe area insets of the view and thus imitating an iphoneX

jakerockland committed:
Added flag propagation tests for shouldIncludeSafeAreaInContentHeight and shouldIncludeSafeAreaInInitialDrawerHeight flags

jakerockland commented:
@yarneo Happy to be contributing here! 😄 

So I tried using https://github.com/material-components/material-components-ios/blob/b8090cb6382910cc4190c7030739e145f3856929/components/FlexibleHeader/tests/unit/supplemental/FlexibleHeaderTopSafeAreaTestsFakeViewController.m as a reference, adding a test in `MDCNavigationDrawerPresentationControllerTests.m` by setting the `MDCBottomDrawerContainerViewController`'s view directly but am not sure if this approach is workable without faking out the `MDCBottomDrawerContainerViewController` itself, which in some way defeats the intent of the test.

This is the approach I tried variations of, which wasn't working for me, though perhaps I was overlooking something silly 😓:

```obj-c
@interface FakeBottomSafeAreaView : UIView
@end

@implementation FakeBottomSafeAreaView

- (UIEdgeInsets)safeAreaInsets {
  UIEdgeInsets safeAreaInsets = [super safeAreaInsets];
  safeAreaInsets.bottom = 34.f;
  return safeAreaInsets;
}

@end

@implementation MDCNavigationDrawerPresentationControllerTests

// ... stuff removed for snippet

- (void)testInitialDrawerHeightBottomSafeArea {
  // Given
  UIViewController *fakeContentViewController = [[UIViewController alloc] init];
  MDCBottomDrawerViewController *fakeBottomDrawer = [[MDCBottomDrawerViewController alloc] init];
  fakeBottomDrawer.contentViewController = fakeContentViewController;
  UIViewController *fakePresentingViewController = [[UIViewController alloc] init];
  MDCBottomDrawerPresentationController *presentationController =
      [[MDCBottomDrawerPresentationController alloc]
          initWithPresentedViewController:fakeBottomDrawer
                 presentingViewController:fakePresentingViewController];
  presentationController.bottomDrawerContainerViewController.view = [[FakeBottomSafeAreaView alloc] init];

  // When
  [presentationController presentationTransitionWillBegin];
  CGFloat height = presentationController.bottomDrawerContainerViewController.maximumInitialDrawerHeight;
  presentationController.shouldIncludeSafeAreaInInitialDrawerHeight = YES;
  CGFloat safeAreaHeight = presentationController.bottomDrawerContainerViewController.maximumInitialDrawerHeight;

  // Then
  XCTAssertTrue(safeAreaHeight > height);
}

// ... stuff removed for snippet

@end
```

A couple tests which are easier to add are just testing the property flag propagation, which may offer sufficient testing, given how simple the logic is for actually adding the bottom inset. 

I've added tests for those to this PR—for both this new flag and the previously existing similar one. 

Thoughts on this?

yarneo commented on components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m:
```diff
@@ -115,4 +115,45 @@ - (void)testSettingShouldAlwaysExpandHeader {
       presentationController.bottomDrawerContainerViewController.shouldAlwaysExpandHeader);
 }
+
+- (void)testShouldIncludeSafeAreaInContentHeight {
```
this test looks good, the bottom one looks like a duplicate? Would we be able to test the propagation from the MDCBottomDrawerViewController to the MDCBottomDrawerPresentationController?

jakerockland replied:
So this test is actually just a bit of a bonus test (could be added in a separate PR but seemed relevant here or I can remove as well) which tests the existing `shouldIncludeSafeAreaInContentHeight` flag for content height taking safe areas into account. 


The test below does the actual testing around `shouldIncludeSafeAreaInInitialDrawerHeight`, which is the newly added but similarly named flag.

Testing though that the flag propagates from MDCBottomDrawerViewController to the MDCBottomDrawerPresentationController sounds good to me.

I'm signing off for today but will revisit this on Monday and I can modify both of these tests to set the flag on the MDCBottomDrawerViewController and verify that it propagates to both the MDCBottomDrawerPresentationController and the container VC?

yarneo replied:
ohh i see now! Perfect! Yes sounds good to me :)

jakerockland replied:
Awesome, thanks @yarneo! I'll make those tweaks to the propagation test then as well as cleaning up the clang-format warning on Monday. Have a great weekend 😄

yarneo commented:
We chatted offline, and the propagation route seems reasonable based on the difficulties of imitating the safe area in the unit test for the drawer container view controller

material-automation commented on components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m:
```diff
@@ -115,4 +115,45 @@ - (void)testSettingShouldAlwaysExpandHeader {
+
+  // Then
+  XCTAssertTrue(
+      presentationController.bottomDrawerContainerViewController.shouldIncludeSafeAreaInContentHeight);
```
clang-format suggested the following change:


```
-  XCTAssertTrue(
-      presentationController.bottomDrawerContainerViewController.shouldIncludeSafeAreaInContentHeight);
+  XCTAssertTrue(presentationController.bottomDrawerContainerViewController
+                    .shouldIncludeSafeAreaInContentHeight);
```

<!-- identifier: clang-format -->
<!-- File: components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m -->

material-automation commented on components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m:
```diff
@@ -115,4 +115,45 @@ - (void)testSettingShouldAlwaysExpandHeader {
+
+  // Then
+  XCTAssertTrue(
+      presentationController.bottomDrawerContainerViewController.shouldIncludeSafeAreaInInitialDrawerHeight);
```
clang-format suggested the following change:


```
-  XCTAssertTrue(
-      presentationController.bottomDrawerContainerViewController.shouldIncludeSafeAreaInInitialDrawerHeight);
+  XCTAssertTrue(presentationController.bottomDrawerContainerViewController
+                    .shouldIncludeSafeAreaInInitialDrawerHeight);
```

<!-- identifier: clang-format -->
<!-- File: components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/NavigationDrawer:snapshot_tests
//components/NavigationDrawer:snapshot_tests_test_bundle
//components/NavigationDrawer:unit_tests
//components/NavigationDrawer:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/NavigationDrawer:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/NavigationDrawer:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/NavigationDrawer:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/NavigationDrawer:unit_tests_environment
//components/NavigationDrawer:unit_tests_environment_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented:
## API diff detected the following changes

### NavigationDrawer

#### MDCBottomDrawerPresentationController

*new* property: `shouldIncludeSafeAreaInInitialDrawerHeight` in `MDCBottomDrawerPresentationController`

#### MDCBottomDrawerViewController

*new* property: `shouldIncludeSafeAreaInInitialDrawerHeight` in `MDCBottomDrawerViewController`

<!-- identifier: apidiff -->

jakerockland committed:
Added additional tests for testing propogation from the MDCBottomDrawerViewController to the MDCBottomDrawerPresentationController

jakerockland commented:
Hey @yarneo, I added some additional tests for testing the propagation from the MDCBottomDrawerViewController to the MDCBottomDrawerPresentationController. I added these both for the existing `shouldIncludeSafeAreaInContentHeight` flag as well as the newly added `shouldIncludeSafeAreaInInitialDrawerHeight` one.

yarneo commented:
Awesome! Looks great! Re-running CI to make sure everything is good to go

yarneo commented:
Thanks a bunch!! Merged the PR
